### PR TITLE
Implemented HRAB algorithm

### PR DESCRIPTION
**Pull request by rpjayasekara**

### CODE REVIEW

**Review by KalanaDananjaya — APPROVED** [review]

File: swarmform/core/clustering_algo/hrab_clustering.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    """
+    n = len(tasks)
+    for i in range(n):
+        for j in range(0, n - i - 1):
+            if tasks[j].get_exec_time() < tasks[j + 1].get_exec_time():
+                tasks[j], tasks[j + 1] = tasks[j + 1], tasks[j]
+    return tasks
+
+
+def get_sum_0f_exec_time(cluster):
```
Comment (AyeshW) [inline swarmform/core/clustering_algo/hrab_clustering.py]: change function name to ```get_sum_of_exec_time``` and refactor

File: swarmform/core/clustering_algo/hrab_clustering.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    Returns:
+        int
+    """
+    exec_sum = 0
+    for task in cluster:
+        exec_sum += task.get_exec_time()
+    return exec_sum
+
+
+def cluster_wf(workflow, cluster_num):
```
Comment (AyeshW) [inline swarmform/core/clustering_algo/hrab_clustering.py]: suggestion: rename function name to 'rac_clustering' as same as 'wpa_clustering'

Comment (AyeshW) [reply swarmform/core/clustering_algo/hrab_clustering.py]: + add a docstring

File: swarmform/core/clustering_algo/hrab_clustering.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+            avg_runtime = get_average_runtime(tasks)
+            avg_cores = get_average_cores(tasks)
+            for task in tasks:
+                candidate_task = get_candidate_cluster(clusters, task, cluster_size, avg_cores, avg_runtime)
+                candidate_task.set_cluster_node(task)
+            for cluster in clusters:
+                update_cluster_info(cluster)
+            update_parent_child_relationships(clusters)
+    workflow.update_links()
+
```
Comment (AyeshW) [inline swarmform/core/clustering_algo/hrab_clustering.py]: return 'workflow' similar to WPA

File: swarmform/core/clustering_algo/hrab_clustering.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            for task in cluster.get_cluster_tasks():
+                update_parent_child_of_a_task(cluster, task)
+
+
+def update_cluster_info(cluster):
+    cluster_info = {}
+    for task in cluster.get_cluster_tasks():
+        cores = task.get_num_cores()
+        runtime = task.get_exec_time()
+        fw_info = {'exec_time': runtime, 'cores': 5}
```
Comment (AyeshW) [inline swarmform/core/clustering_algo/hrab_clustering.py]: replace static value

File: swarmform/core/clustering_algo/hrab_clustering.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+filename = '/Users/randika/Documents/FYP/SwarmForm/swarmform/examples/swarmpad_examples/Montage_25-1588016685/Montage_25.yaml'
+# create the Firework consisting of a custom "Addition" task
+unclustered_sf = SwarmFlow.from_file(filename)
+
+# store workflow
+swarmpad.add_sf(unclustered_sf)
+sf = swarmpad.get_sf_by_id(unclustered_sf.sf_id)
+sf_dag = DAG(sf)
+
+cluster_wf(sf_dag, 3)
```
Comment (AyeshW) [inline swarmform/core/clustering_algo/hrab_clustering.py]: remove at the end

**Review by AyeshW — CHANGES_REQUESTED** [review]
Let's try to return outputs as same as WPA for minimizing further issues

**Review by AyeshW — APPROVED** [review]
Ready to be merged

**Pull request merged**